http/src/main/resources/jupyter/gce-init.sh

```diff
@@ -0,0 +1,528 @@
+          retry 3 gsutil -h "x-goog-meta-passed":"true" cp start_output.txt ${JUPYTER_START_USER_SCRIPT_OUTPUT_URI}
+        fi
+      fi
+
+      STEP_TIMINGS+=($(date +%s))
+
+      # Install lab extensions
+      # Note: lab extensions need to installed as jupyter user, not root
+      if [ ! -z "${JUPYTER_LAB_EXTENSIONS}" ] ; then
+        for ext in ${JUPYTER_LAB_EXTENSIONS}
+        do
+          log 'Installing JupyterLab extension [$ext]...'
+          pwd
+          if [[ $ext == 'gs://'* ]]; then
+            gsutil cp -r $ext /etc
+            JUPYTER_EXTENSION_ARCHIVE=`basename $ext`
+            docker cp /etc/${JUPYTER_EXTENSION_ARCHIVE} ${JUPYTER_SERVER_NAME}:${JUPYTER_HOME}/${JUPYTER_EXTENSION_ARCHIVE}
+            retry 3 docker exec ${JUPYTER_SERVER_NAME} ${JUPYTER_SCRIPTS}/extension/jupyter_install_lab_extension.sh ${JUPYTER_HOME}/${JUPYTER_EXTENSION_ARCHIVE}
+          elif [[ $ext == 'http://'* || $ext == 'https://'* ]]; then
+            JUPYTER_EXTENSION_FILE=`basename $ext`
+            curl $ext -o /etc/${JUPYTER_EXTENSION_FILE}
+            docker cp /etc/${JUPYTER_EXTENSION_FILE} ${JUPYTER_SERVER_NAME}:${JUPYTER_HOME}/${JUPYTER_EXTENSION_FILE}
+            retry 3 docker exec ${JUPYTER_SERVER_NAME} ${JUPYTER_SCRIPTS}/extension/jupyter_install_lab_extension.sh ${JUPYTER_HOME}/${JUPYTER_EXTENSION_FILE}
+          else
+            retry 3 docker exec ${JUPYTER_SERVER_NAME} ${JUPYTER_SCRIPTS}/extension/jupyter_install_lab_extension.sh $ext
+          fi
+        done
+      fi
+
+      STEP_TIMINGS+=($(date +%s))
+
+      # fix for https://broadworkbench.atlassian.net/browse/IA-1453
+      # TODO: remove this when we stop supporting the legacy docker image
+      if [ ! -z ${WELDER_DOCKER_IMAGE} ] && [ "${WELDER_ENABLED}" == "true" ] ; then
+        retry 3 docker exec -u root ${JUPYTER_SERVER_NAME} sed -i -e 's/export WORKSPACE_NAME=.*/export WORKSPACE_NAME="$(basename "$(dirname "$(pwd)")")"/' ${JUPYTER_HOME}/scripts/kernel/kernel_bootstrap.sh
+      fi
+
+      STEP_TIMINGS+=($(date +%s))
+
+      log 'Starting Jupyter Notebook...'
+      retry 3 docker exec -d ${JUPYTER_SERVER_NAME} ${JUPYTER_SCRIPTS}/run-jupyter.sh ${NOTEBOOKS_DIR}
+
+      STEP_TIMINGS+=($(date +%s))
+    fi
+
+    # Remove any unneeded cached images to save disk space.
+    # Do this asynchronously so it doesn't hold up cluster creation
+    log 'Pruning docker images...'
+    docker image prune -a -f &
+elif [[ "${GCE_OPERATION}" == 'restarting' ]]; then
+  export DEPLOY_WELDER=$(deployWelder)
+  export UPDATE_WELDER=$(updateWelder)
+  export DISABLE_DELOCALIZATION=$(disableDelocalization)
+
+  # Sometimes we want to update Welder without having to delete and recreate a cluster
+  if [ "$UPDATE_WELDER" == "true" ] ; then
+      gcloud auth configure-docker
+      docker-compose -f /etc/welder-docker-compose.yaml stop
+      docker-compose -f /etc/welder-docker-compose.yaml rm -f
+      docker-compose -f /etc/welder-docker-compose.yaml up -d
```

Comment:
Maybe this should also have a `docker-compose pull` with a retry

Reply:
You mean, retry `docker-compose up`?

Reply:
If there's no `pull` then `up` does the pull. But I think it's better to have an explicit `pull` with retries.